src/transformStateWithClones.js

```diff
@@ -6,8 +6,33 @@
  *
  * @return {Object[]}
  */
-function transformStateWithClones(state, actions) {
-  // write code here
+function transformStateWithClones(initialState, actions) {
+  const stateHistory = [];
+  let currentState = { ...initialState };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        currentState = {};
+        break;
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          if (currentState.hasOwnProperty(key)) {
+            delete currentState[key];
+          }
+        }
+        break;
+      default:
+        throw new Error(`Unknown action type: ${action.type}`);
```

Comment:
Consider moving the operation of cloning the `currentState` outside of the switch statement since it's repeated in every case. This will adhere to the DRY (Don't Repeat Yourself) principle.